src/sage/schemes/affine/affine_morphism.py

```diff
@@ -640,31 +647,31 @@ def as_dynamical_system(self):
             sage: A.<x,y,z> = AffineSpace(ZZ, 3)
             sage: H = End(A)
             sage: f = H([x^2, y^2, z^2])
-            sage: type(f.as_dynamical_system())
+            sage: type(f.as_dynamical_system())                                         # needs sage.schemes
             <class 'sage.dynamics.arithmetic_dynamics.affine_ds.DynamicalSystem_affine'>
```

Comment:
Is `sage.schemes` necessary for affine morphisms?

Reply:
No. `schemes.{generic,affine,projective,product_projective}` are shipped already by sagemath-categories.

Reply:
Then why is there `# needs sage.schemes` here?

Reply:
Ah, I see.

But then it looks strange to me that `schemes.{generic,affine,projective,product_projective}` are shipped by sagemath-categories not by `sagemath.schemes`.  But I guess there are technical reasons.

Personally I think `sage.dynamics` should be separate from `sage.schemes`, but depend on `sage.schemes`. I don't know how they are organized now, but at least `# needs sage.schemes` here looks odd to me.

Reply:
It's for the same reasons that sagemath-categories also ships basic modules from `sage.combinat`, or `sage.rings.padics`: The basic modules are needed somewhere else where the technical parts of sagemath-schemes (in particular the dependency on Singular) are not needed. This includes `isinstance` testing. Specifically, elements of the affine/projective/... spaces are handled both in `sage.rings` and `sage.matrix`

Reply:
A feature `sage.dynamics` would not be very clean because this package includes various technically unrelated subpackages. A feature `sage.dynamics.arithmetic_dynamics` (named after the package) could make sense, but I'll leave it to the experts whether introducing this feature is meaningful. From the modularization point of view, nothing interesting is happening there.

Reply:
I didn't understand the situation when I made the comment above.

Mathematical content (ambient spaces) of `sage.schemes` are already in `sage.categories`. So the feature `sage.schemes` seems to essentially cover curves. It would make more sense to rename `sage.schemes` to `sage.schemes.curves`. Then the above tag  to `type(f.as_dynamical_system())` would look reasonable. But I don't suggest doing this as `sage.schemes` may cover more things.

Reply:
Perhaps I thought wrong. It is morphisms that dynamical systems are using.

Reply:
Anyway, ok.